Assets/Documentation/man/man3/MidiInput.3
.TH "MidiInput" 3 "Version 1.0.0" "KiBoard GDD & Technical Documentation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MidiInput
.SH SYNOPSIS
.br
.PP
.PP
Inherits MonoBehaviour\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBLoadScenePreview\fP (string sceneName)"
.br
.RI "Loads a preview of selected song into the \fBcurrentPreview\fP Scene variable, and begins playing the songs preview\&. "
.ti -1c
.RI "void \fBLoadSongFromCurrentSettings\fP (bool isPreview=false)"
.br
.RI "Loads the selected song for gameplay\&. "
.ti -1c
.RI "void \fBHookMidiDevice\fP ()"
.br
.RI "Hooks midi devices to be used for input\&. "
.ti -1c
.RI "void \fBUnHookMidiDevice\fP ()"
.br
.RI "unhooks midi devices to remove their input\&. "
.ti -1c
.RI "\fBNoteEventDataWrapper\fP \fBGetNoteEventWrapperFromSelectedSong\fP ()"
.br
.ti -1c
.RI "void \fBStartSongCoroutine\fP ()"
.br
.RI "called to begin playing the song\&. "
.ti -1c
.RI "IEnumerator \fBStartSong\fP (bool isPreview=false)"
.br
.RI "Starts playing the loaded song using the current \fBFileGroup\fP's mp3 file field\&. "
.ti -1c
.RI "IEnumerator \fBStartSong\fP (string mp3Path, bool isPreview=false)"
.br
.RI "Starts playing the loaded song from a given file path\&. "
.ti -1c
.RI "IEnumerator \fBStartSong\fP (List< \fBNoteEventInfo\fP > loadEvents, bool isPreview=false)"
.br
.RI "Starts playing a specified list of \fBNoteEventInfo\fP "
.ti -1c
.RI "void \fBStopSong\fP ()"
.br
.RI "Stops the currently playing song\&. "
.ti -1c
.RI "void \fBOnNoteSuccess\fP (ref string score, int note, float timing, \fBNoteEventInfo\fP storedNote)"
.br
.RI "Handles successful note presses\&. "
.ti -1c
.RI "bool \fBIsNoteCorrect\fP (int noteNumber, float timing, \fBNoteEventInfo\fP storedNote)"
.br
.RI "Check if note is was played at a correct timing\&. "
.ti -1c
.RI "void \fBRetryCurrentSong\fP ()"
.br
.RI "Reloads the currently playing song\&. "
.ti -1c
.RI "bool \fBGetAnyNoteActive\fP ()"
.br
.RI "Checks for pressed keys in the 8-key keyboard layout\&. "
.ti -1c
.RI "string \fBGetTimingScore\fP (float timing)"
.br
.RI "Gets the timing score based on the timing difference\&. "
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "List< \fBNoteEventInfo\fP > \fBstoredNoteEvents\fP"
.br
.RI "Currently loaded notes for the selected song\&. "
.ti -1c
.RI "Coroutine \fBPrepareNotesCoroutine\fP"
.br
.RI "Reference to the \fBGameManager\&.PrepareNotes(float, List<NoteEventInfo>, bool)\fP prepareNotes Coroutine\&. "
.ti -1c
.RI "bool \fBisPedalPressed\fP"
.br
.RI "Is the pedal pressed\&. "
.ti -1c
.RI "bool[] \fBenabledKeys\fP = new bool[88]"
.br
.RI "array of bools corresponding to currently activated keys\&. "
.ti -1c
.RI "bool \fBtakeInput\fP = true"
.br
.RI "Should the game take piano input\&. "
.ti -1c
.RI "bool \fBinGame\fP = false"
.br
.RI "Are we in the game scene\&. "
.ti -1c
.RI "bool \fBisMidiHooked\fP"
.br
.RI "are midi devices hooked?
.br
 \fBHookMidiDevice\fP and \fBUnHookMidiDevice\fP\&. "
.ti -1c
.RI "RenderTexture \fBrenderTexture\fP"
.br
.RI "Render texture used for previewing scenes in Song Selection\&. "
.ti -1c
.RI "Scene \fBcurrentPreview\fP"
.br
.RI "Scene used for previewing songs in Song Selection\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static \fBMidiInput\fP \fBinstance\fP"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBAwake\fP ()"
.br
.ti -1c
.RI "void \fBUpdate\fP ()"
.br
.ti -1c
.RI "void \fBPedalStateChanged\fP (MidiChannel channel, int knobNumber, float knobValue)"
.br
.RI "Handles MIDI sustain pedal state changes\&. "
.ti -1c
.RI "void \fBNoteOn\fP (MidiChannel channel, int note, float velocity)"
.br
.RI "Handles MIDI note on events\&. "
.ti -1c
.RI "void \fBNoteOff\fP (MidiChannel channel, int note)"
.br
.RI "Handles MIDI note off events\&. "
.ti -1c
.RI "void \fBCheckNotesKeyboard12\fP ()"
.br
.RI "Checks for pressed keys in the 12-key keyboard layout\&. "
.ti -1c
.RI "float \fBGetTimeDifference\fP (float storedTiming)"
.br
.RI "Calculates the time difference between the stored timing and the current song time\&. "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "GameObject \fBimagePrefab\fP"
.br
.RI "prefab object with a RawImage component to assign \fBrenderTexture\fP to\&. "
.ti -1c
.RI "Dictionary< KeyCode, int > \fBkeyboard12\fP"
.br
.RI "Dictionary mapping for Computer keyboard input to note input\&. "
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "void MidiInput\&.Awake ()\fR [private]\fP"

.nf
82     {
83         if (instance == null)
84         {
85             instance = this;
86             DontDestroyOnLoad(gameObject);
87         }
88         else { Destroy(gameObject); }
89     }
.PP
.fi

.SS "void MidiInput\&.CheckNotesKeyboard12 ()\fR [private]\fP"

.PP
Checks for pressed keys in the 12-key keyboard layout\&. 
.nf
481     {
482         // Check for computer keyboard key presses and releases
483         foreach (var keyValuePair in keyboard12)
484         {
485             KeyCode key = keyValuePair\&.Key;
486             int noteNumber = keyValuePair\&.Value;
487 
488             if (Input\&.GetKeyDown(key))
489             {
490                 // Simulate Note On event for the pressed key
491                 NoteOn(MidiChannel\&.All, noteNumber, 1\&.0f);
492             }
493 
494             if (Input\&.GetKeyUp(key))
495             {
496                 // Simulate Note Off event for the released key
497                 NoteOff(MidiChannel\&.All, noteNumber);
498             }
499         }
500     }
.PP
.fi

.PP
Referenced by \fBUpdate()\fP\&.
.SS "bool MidiInput\&.GetAnyNoteActive ()"

.PP
Checks for pressed keys in the 8-key keyboard layout\&. Checks if any note is currently active (pressed)\&.
.PP
\fBReturns\fP
.RS 4
True if any note is active, otherwise false\&.
.RE
.PP

.nf
511     {
512         foreach (bool b in enabledKeys)
513         {
514             if (b)
515             {
516                 return true;
517             }
518         }
519         return false;
520     }
.PP
.fi

.SS "\fBNoteEventDataWrapper\fP MidiInput\&.GetNoteEventWrapperFromSelectedSong ()"

.PP
\fBReturns\fP
.RS 4
\fBNoteEventDataWrapper\fP loaded from the currently selected song
.RE
.PP

.nf
310     {
311         NoteEventDataWrapper data = MidiReadFile\&.GetNoteEventsFromFilePath(GameSettings\&.currentSongPath);
312         if (data == null) { Debug\&.Log("Error Reading data from path: " + GameSettings\&.currentSongPath); return null; }
313         storedNoteEvents = data\&.NoteEvents;
314         return data;
315     }
.PP
.fi

.SS "float MidiInput\&.GetTimeDifference (float storedTiming)\fR [private]\fP"

.PP
Calculates the time difference between the stored timing and the current song time\&. 
.PP
\fBParameters\fP
.RS 4
\fIstoredTiming\fP The stored timing value\&.
.RE
.PP
\fBReturns\fP
.RS 4
The time difference\&.
.RE
.PP

.nf
528     {
529         // Adjust this threshold based on your timing accuracy requirements
530         // float timingThreshold = currentSettings\&.timeInterval;
531 
532         // Check if the received timing is within a certain threshold of the stored timing
533         //0\&.125f leniancy
534         return Mathf\&.Abs( (GameManager\&.instance\&.songTime + PlayerSettings\&.inputDelay)  \- 0\&.125f \- storedTiming);
535     }
.PP
.fi

.PP
Referenced by \fBNoteOn()\fP\&.
.SS "string MidiInput\&.GetTimingScore (float timing)"

.PP
Gets the timing score based on the timing difference\&. 
.PP
\fBParameters\fP
.RS 4
\fItiming\fP The timing difference\&.
.RE
.PP
\fBReturns\fP
.RS 4
The timing score (e\&.g\&., 'Perfect', 'Good', 'Okay', 'Miss')\&.
.RE
.PP

.nf
542     {
543         if (timing < 0\&.1f) { return "Perfect"; }
544         else if (timing < 0\&.175f) { return "Good"; }
545         else if (timing < 0\&.275f) { return "Okay"; }
546         else { return "Miss"; }
547     }
.PP
.fi

.PP
Referenced by \fBNoteOn()\fP, and \fBOnNoteSuccess()\fP\&.
.SS "void MidiInput\&.HookMidiDevice ()"

.PP
Hooks midi devices to be used for input\&. 
.nf
269     {
270         try
271         {
272             if (isMidiHooked) { Debug\&.Log("Midi Already Hooked"); return; }
273             MidiMaster\&.noteOnDelegate += NoteOn;
274             MidiMaster\&.noteOffDelegate += NoteOff;
275             MidiMaster\&.knobDelegate += PedalStateChanged;
276             isMidiHooked = true;
277         }
278         catch
279         {
280             isMidiHooked = false;
281             Debug\&.Log("Error Hooking midi device");
282         }
283     }
.PP
.fi

.PP
Referenced by \fBLoadSongFromCurrentSettings()\fP\&.
.SS "bool MidiInput\&.IsNoteCorrect (int noteNumber, float timing, \fBNoteEventInfo\fP storedNote)"

.PP
Check if note is was played at a correct timing\&. 
.PP
\fBParameters\fP
.RS 4
\fInoteNumber\fP Number of the note played\&.
.br
\fItiming\fP What time the note was hit\&.
.br
\fIstoredNote\fP Note that is being compared to\&.
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.nf
445     {
446         if (GameSettings\&.usePiano) { return noteNumber == storedNote\&.noteNumber && !storedNote\&.triggered && timing < 0\&.5f; }
447         return noteNumber % 12 == storedNote\&.noteNumber % 12 && !storedNote\&.triggered && timing < 0\&.5f;
448 
449     }
.PP
.fi

.PP
Referenced by \fBNoteOn()\fP\&.
.SS "void MidiInput\&.LoadScenePreview (string sceneName)"

.PP
Loads a preview of selected song into the \fBcurrentPreview\fP Scene variable, and begins playing the songs preview\&. 
.PP
\fBParameters\fP
.RS 4
\fIsceneName\fP 
.RE
.PP

.nf
128     {
129 
130         if (currentPreview\&.isLoaded)
131         {
132             SceneManager\&.UnloadSceneAsync(currentPreview);
133         }
134         AsyncOperation loadOperation = SceneManager\&.LoadSceneAsync(sceneName, LoadSceneMode\&.Additive);
135         loadOperation\&.completed += (operation) =>
136         {
137             currentPreview = SceneManager\&.GetSceneByName(sceneName); // Assign the loaded scene to currentPreview
138             OnSceneLoaded(operation);
139         };
140 
141 
142         void OnSceneLoaded(AsyncOperation asyncOperation)
143         {
144 
145             Scene previewScene = SceneManager\&.GetSceneByName(sceneName);
146 
147 
148             // Iterate through the root game objects of the scene
149             foreach (GameObject rootObject in previewScene\&.GetRootGameObjects())
150             {
151                 if (rootObject\&.TryGetComponent(out TransitionManager transition))
152                 {
153                     transition\&.GetComponent<Canvas>()\&.enabled = false;
154                 }
155                 if (rootObject\&.TryGetComponent(out Canvas canvas))
156                 {
157                     canvas\&.enabled = false;
158                 }
159                 // Assign the object to the "PreviewLayer"
160                 AssignToPreviewLayer(rootObject);
161 
162                 // Check if the object has an EventSystem component and destroy it
163                 if (rootObject\&.TryGetComponent(out EventSystem eventSystem))
164                 {
165                     Destroy(eventSystem\&.gameObject);
166                 }
167 
168                 // Find the camera in the preview scene
169                 Camera camera = rootObject\&.GetComponentInChildren<Camera>();
170                 if (camera != null)
171                 {
172 
173                     // Set the camera to render to the specified RenderTexture
174                     camera\&.cullingMask = 1 << LayerMask\&.NameToLayer("PreviewLayer");
175                     camera\&.targetTexture = renderTexture;
176                 }
177             }
178 
179             // Create a new UI Image object
180             GameObject imageObject = Instantiate(imagePrefab, UiHolder\&.instance\&.transform);
181             RawImage image = imageObject\&.GetComponentInChildren<RawImage>(); // CAN ONLY DO THIS BECAUSE IT IS RAWIMAGE, SINCE I USE MANY NORMAL IMAGE TYPES\&.
182 
183 
184             // Check if RawImage component exists
185             if (image != null)
186             {
187                 // Assign the RenderTexture to the RawImage component's texture
188                 image\&.texture = renderTexture;
189                 if (UiHolder\&.instance\&.scenePreview != null) { Destroy(UiHolder\&.instance\&.scenePreview); }
190                 UiHolder\&.instance\&.scenePreview = imageObject;
191             }
192             else
193             {
194                 Debug\&.LogError("RawImage component not found on the instantiated image prefab\&.");
195             }
196             asyncOperation\&.completed \-= OnSceneLoaded;
197 
198         }
199 
200         // Method to assign an object and its children to the "PreviewLayer"
201         void AssignToPreviewLayer(GameObject obj)
202         {
203             // Assign the object to the "PreviewLayer"
204             obj\&.layer = LayerMask\&.NameToLayer("PreviewLayer");
205 
206             // Recursively assign children to the "PreviewLayer"
207             foreach (Transform child in obj\&.transform)
208             {
209                 AssignToPreviewLayer(child\&.gameObject);
210             }
211         }
212     }
.PP
.fi

.PP
Referenced by \fBLoadSongFromCurrentSettings()\fP\&.
.SS "void MidiInput\&.LoadSongFromCurrentSettings (bool isPreview = \fRfalse\fP)"

.PP
Loads the selected song for gameplay\&. 
.nf
221     {
222         MP3Handler\&.instance\&.StopMusic();
223         string gameMode;
224         if (KiboardDebug\&.isMidiConnected && GameSettings\&.usePiano) { gameMode = "GameScene88"; } else { gameMode = "GameScene12"; }
225 
226         if (isPreview)
227         {
228             LoadScenePreview(gameMode);
229             NoteEventDataWrapper data = MidiReadFile\&.GetNoteEventsFromFilePath(GameSettings\&.currentSongPath);
230             GameSettings\&.bpm = data\&.BPM;
231             storedNoteEvents = data\&.NoteEvents;
232             takeInput = false;
233             inGame = false;
234             StartCoroutine(StartSong(true));
235         }
236         else
237         {
238 
239             try
240             {
241                 TransitionManager\&.instance\&.LoadNewScene(gameMode);
242             }
243 
244             catch(Exception e)
245             {
246                 Debug\&.Log(e);
247                 SceneManager\&.LoadScene(gameMode);
248             }
249 
250             if (GameSettings\&.usePiano) { HookMidiDevice(); } else { UnHookMidiDevice(); }
251             NoteEventDataWrapper data = MidiReadFile\&.GetNoteEventsFromFilePath(GameSettings\&.currentSongPath);
252             GameSettings\&.bpm = GameSettings\&.bpm == 0 ? data\&.BPM : GameSettings\&.bpm;
253 
254             storedNoteEvents = data\&.NoteEvents;
255             takeInput = true;
256             inGame = true;
257             if (PrepareNotesCoroutine != null) StopCoroutine(PrepareNotesCoroutine);
258             GameManager\&.instance\&.StopReadiedNotes();
259             StartCoroutine(StartSong());
260         }
261 
262 
263 
264     }
.PP
.fi

.PP
Referenced by \fBRetryCurrentSong()\fP\&.
.SS "void MidiInput\&.NoteOff (MidiChannel channel, int note)\fR [private]\fP"

.PP
Handles MIDI note off events\&. 
.PP
\fBParameters\fP
.RS 4
\fIchannel\fP The MIDI channel\&.
.br
\fInote\fP The MIDI note number\&.
.RE
.PP

.nf
458     {
459         if (!takeInput || isPedalPressed) { return; }
460         if (inGame)
461         {
462             SpawnPiano\&.instance\&.UpdateKeyColours(note \- 21, false);
463             Replay\&.UpdateReplay(note, GameManager\&.instance\&.songTime);
464         }
465         enabledKeys[note \- 21] = false;
466 
467     }
.PP
.fi

.PP
Referenced by \fBCheckNotesKeyboard12()\fP, \fBHookMidiDevice()\fP, and \fBUnHookMidiDevice()\fP\&.

Assets/Scripts/MidiInput.cs
using MidiJack;
using System;
using System.Collections;
using System.Collections.Generic;
using TMPro;
using UnityEngine;
using UnityEngine.EventSystems;
using UnityEngine.Rendering.Universal;
using UnityEngine.SceneManagement;
using UnityEngine.UI;
using static Cinemachine.CinemachineTriggerAction.ActionSettings;

public class MidiInput : MonoBehaviour
{
    public static MidiInput instance;
    /// <summary>
    /// Currently loaded notes for the selected song.
    /// </summary>
    public List<NoteEventInfo> storedNoteEvents;
    /// <summary>
    /// Reference to the <see cref="GameManager.PrepareNotes(float, List{NoteEventInfo}, bool)"/> prepareNotes Coroutine 
    /// </summary>
    public Coroutine PrepareNotesCoroutine;
    /// <summary>
    /// Is the pedal pressed.
    /// </summary>
    public bool isPedalPressed;
    /// <summary>
    /// array of bools corresponding to currently activated keys.
    /// </summary>
    public bool[] enabledKeys = new bool[88];
    /// <summary>
    /// Should the game take piano input.
    /// </summary>
    public bool takeInput = true;
    /// <summary>
    /// Are we in the game scene.
    /// </summary>
    public bool inGame = false;

    /// <summary>
    /// are midi devices hooked?<br/>
    /// <see cref="HookMidiDevice"/> and <see cref="UnHookMidiDevice"/>.
    /// </summary>
    public bool isMidiHooked;

    /// <summary>
    /// Render texture used for previewing scenes in Song Selection.
    /// </summary>
    public RenderTexture renderTexture;
    /// <summary>
    /// keyPrefab object with a RawImage component to assign <see cref="renderTexture"/> to.
    /// </summary>
    [SerializeField] GameObject imagePrefab;
    
    /// <summary>
    /// Dictionary mapping for Computer keyboard input to note input.
    /// </summary>
    Dictionary<KeyCode, int> keyboard12 = new Dictionary<KeyCode, int>
    {
        { KeyCode.A, 48 },
        { KeyCode.W, 49 },
        { KeyCode.S, 50 },
        { KeyCode.E, 51 },
        { KeyCode.D, 52 },
        { KeyCode.F, 53 },
        { KeyCode.T, 54 },
        { KeyCode.G, 55 },
        { KeyCode.Y, 56 },
        { KeyCode.H, 57 },
        { KeyCode.U, 58 },
        { KeyCode.J, 59 },
    };

    /// <summary>
    /// Scene used for previewing songs in Song Selection.
    /// </summary>
    public Scene currentPreview;


    private void Awake()
    {
        if (instance == null)
        {
            instance = this;
            DontDestroyOnLoad(gameObject);
        }
        else { Destroy(gameObject); }
    }

    private void Update()
    {
        if (Input.GetKeyDown(KeyCode.Escape))
        {
            StopSong();
        }
        CheckNotesKeyboard12();
        //CheckNotesKeyboard8();

    }

    /// <summary>
    /// Handles MIDI sustain pedal state changes.
    /// </summary>
    /// <param name="channel">The MIDI channel.</param>
    /// <param name="knobNumber">The knob number.</param>
    /// <param name="knobValue">The knob value.</param>
    void PedalStateChanged(MidiChannel channel, int knobNumber, float knobValue)
    {
        if (knobNumber == 64 && knobValue > 0.5f)
        {
            isPedalPressed = true;

        }
        else
        {
            isPedalPressed = false;
            try { SpawnPiano.instance.ClearAllKeyColours(); }
            catch { }
        }
    }

 /// <summary>
 /// Loads a preview of selected song into the <see cref="currentPreview"/> Scene variable, and begins playing the songs preview.
 /// </summary>
 /// <param name="sceneName"></param>
    public void LoadScenePreview(string sceneName)
    {

        if (currentPreview.isLoaded)
        {
            SceneManager.UnloadSceneAsync(currentPreview);
        }
        AsyncOperation loadOperation = SceneManager.LoadSceneAsync(sceneName, LoadSceneMode.Additive);
        loadOperation.completed += (operation) =>
        {
            currentPreview = SceneManager.GetSceneByName(sceneName); // Assign the loaded scene to currentPreview
            OnSceneLoaded(operation);
        };


        void OnSceneLoaded(AsyncOperation asyncOperation)
        {

            Scene previewScene = SceneManager.GetSceneByName(sceneName);


            // Iterate through the root game objects of the scene
            foreach (GameObject rootObject in previewScene.GetRootGameObjects())
            {
                if (rootObject.TryGetComponent(out TransitionManager transition))
                {
                    transition.GetComponent<Canvas>().enabled = false;
                }
                if (rootObject.TryGetComponent(out Canvas canvas))
                {
                    canvas.enabled = false;
                }
                // Assign the object to the "PreviewLayer"
                AssignToPreviewLayer(rootObject);

                // Check if the object has an EventSystem component and destroy it
                if (rootObject.TryGetComponent(out EventSystem eventSystem))
                {
                    Destroy(eventSystem.gameObject);
                }

                // Find the camera in the preview scene
                Camera camera = rootObject.GetComponentInChildren<Camera>();
                if (camera != null)
                {

                    // Set the camera to render to the specified RenderTexture
                    camera.cullingMask = 1 << LayerMask.NameToLayer("PreviewLayer");
                    camera.targetTexture = renderTexture;
                }
            }

            // Create a new UI Image object
            GameObject imageObject = Instantiate(imagePrefab, UiHolder.instance.transform);
            RawImage image = imageObject.GetComponentInChildren<RawImage>(); // CAN ONLY DO THIS BECAUSE IT IS RAWIMAGE, SINCE I USE MANY NORMAL IMAGE TYPES.


            // Check if RawImage component exists
            if (image != null)
            {
                // Assign the RenderTexture to the RawImage component's texture
                image.texture = renderTexture;
                if (UiHolder.instance.scenePreview != null) { Destroy(UiHolder.instance.scenePreview); }
                UiHolder.instance.scenePreview = imageObject;
            }
            else
            {
                Debug.LogError("RawImage component not found on the instantiated image prefab.");
            }
            asyncOperation.completed -= OnSceneLoaded;

        }

        // Method to assign an object and its children to the "PreviewLayer"
        void AssignToPreviewLayer(GameObject obj)
        {
            // Assign the object to the "PreviewLayer"
            obj.layer = LayerMask.NameToLayer("PreviewLayer");

            // Recursively assign children to the "PreviewLayer"
            foreach (Transform child in obj.transform)
            {
                AssignToPreviewLayer(child.gameObject);
            }
        }
    }



    /// <summary>
    /// Loads the selected song for gameplay.
    /// </summary>

    public void LoadSongFromCurrentSettings(bool isPreview = false)
    {
        MP3Handler.instance.StopMusic();
        string gameMode;
        if (KiboardDebug.isMidiConnected && GameSettings.usePiano) { gameMode = "GameScene88"; } else { gameMode = "GameScene12"; }

        if (isPreview)
        {
            LoadScenePreview(gameMode);
            NoteEventDataWrapper data = MidiReadFile.GetNoteEventsFromFilePath(GameSettings.currentSongPath);
            GameSettings.bpm = data.BPM;
            storedNoteEvents = data.NoteEvents;
            takeInput = false;
            inGame = false;
            StartCoroutine(StartSong(true));
        }
        else
        {

            try
            {
                TransitionManager.instance.LoadNewScene(gameMode);
            }

            catch(Exception e)
            {
                Debug.Log(e);
                SceneManager.LoadScene(gameMode);
            }

            if (GameSettings.usePiano) { HookMidiDevice(); } else { UnHookMidiDevice(); }
            NoteEventDataWrapper data = MidiReadFile.GetNoteEventsFromFilePath(GameSettings.currentSongPath);
            GameSettings.bpm = GameSettings.bpm == 0 ? data.BPM : GameSettings.bpm;

            storedNoteEvents = data.NoteEvents;
            takeInput = true;
            inGame = true;
            if (PrepareNotesCoroutine != null) StopCoroutine(PrepareNotesCoroutine);
            GameManager.instance.StopReadiedNotes();
            StartCoroutine(StartSong());
        }



    }
    /// <summary>
    /// Hooks midi devices to be used for input.
    /// </summary>
    public void HookMidiDevice()
    {
        try
        {
            if (isMidiHooked) { Debug.Log("Midi Already Hooked"); return; }
            MidiMaster.noteOnDelegate += NoteOn;
            MidiMaster.noteOffDelegate += NoteOff;
            MidiMaster.knobDelegate += PedalStateChanged;
            isMidiHooked = true;
        }
        catch
        {
            isMidiHooked = false;
            Debug.Log("Error Hooking midi device");
        }
    }
    /// <summary>
    /// unhooks midi devices to remove their input.
    /// </summary>
    public void UnHookMidiDevice()
    {
        
        try
        {
            MidiMaster.noteOnDelegate -= NoteOn;
            MidiMaster.noteOffDelegate -= NoteOff;
            MidiMaster.knobDelegate -= PedalStateChanged;
            isMidiHooked = false;
        }
        catch
        {
            isMidiHooked = false;
            Debug.Log("Error Unhooking midi device");
        }
    }


    /// <summary>
    /// 
    /// </summary>
    /// <returns><see cref="NoteEventDataWrapper"/> loaded from the currently selected song</returns>
    public NoteEventDataWrapper GetNoteEventWrapperFromSelectedSong()
    {
        NoteEventDataWrapper data = MidiReadFile.GetNoteEventsFromFilePath(GameSettings.currentSongPath);
        if (data == null) { Debug.Log("Error Reading data from path: " + GameSettings.currentSongPath); return null; }
        storedNoteEvents = data.NoteEvents;
        return data;
    }
    /// <summary>
    /// called to begin playing the song.
    /// </summary>
    public void StartSongCoroutine()
    {
        StartCoroutine(StartSong());
    }
    /// <summary>
    /// Starts playing the loaded song using the current <see cref="FileGroup"/>'s mp3 file field.
    /// </summary>
    /// <returns></returns>
    public IEnumerator StartSong(bool isPreview = false)
    {

        GameManager.instance.currentSongScore.ClearScore();
        GameManager.instance.combo.ClearCombo();
        yield return PrepareNotesCoroutine = StartCoroutine(GameManager.instance.PrepareNotes(GameSettings.bpm, storedNoteEvents, isPreview));
        StartCoroutine(MP3Handler.instance.PlaySong(SongSelection.GetUnderscoreSubstring(GameSettings.currentFileGroup.Mp3File)));
    }
    /// <summary>
    /// Starts playing the loaded song from a given file path.
    /// </summary>
    /// <returns></returns>
    public IEnumerator StartSong(string mp3Path, bool isPreview = false)
    {

        GameManager.instance.currentSongScore.ClearScore();
        GameManager.instance.combo.ClearCombo();
        yield return PrepareNotesCoroutine = StartCoroutine(GameManager.instance.PrepareNotes(GameSettings.bpm, storedNoteEvents, isPreview));
        StartCoroutine(MP3Handler.instance.PlaySong(mp3Path));
    }
    /// <summary>
    /// Starts playing a specified list of <see cref="NoteEventInfo"/>
    /// </summary>
    /// <param name="loadEvents">The list of note events to play.</param>
    /// <returns></returns>
    public IEnumerator StartSong(List<NoteEventInfo> loadEvents, bool isPreview = false)
    {
        GameManager.instance.currentSongScore.ClearScore();
        var bpm = GameSettings.bpm;
        loadEvents.ForEach(noteEvent => noteEvent.noteNumber += 20); // i - for the fucking life of me- cannot figure out why directly processing the midi files makes the note numbers
                                                                     // 20 higher, but i have to do this to match that with the song editor.
        yield return PrepareNotesCoroutine = StartCoroutine(GameManager.instance.PrepareNotes(GameSettings.bpm, storedNoteEvents, isPreview));

        StartCoroutine(MP3Handler.instance.PlaySong(GameSettings.currentSongPath));

    }

    /// <summary>
    /// Stops the currently playing song.
    /// </summary>
    public void StopSong()
    {
        if (GameManager.instance.isCurSongPreview) { return; }
        GameManager.instance.StopSong();
        StopCoroutine(PrepareNotesCoroutine);
        MP3Handler.instance.StopMusic();

        GameSettings.ResetSettings(false);
        GameManager.instance.ReturnToSongSelection();


    }

    /// <summary>
    /// Handles MIDI note on events.
    /// </summary>
    /// <param name="channel">The MIDI channel.</param>
    /// <param name="note">The MIDI note number.</param>
    /// <param name="velocity">The note velocity.</param>
    void NoteOn(MidiChannel channel, int note, float velocity)
    {
        if (!takeInput) { return; }
        // Check if the received note and timing match any stored events
        if (inGame)
        {
            string score = "";
            float timing = Mathf.Infinity;
            foreach (NoteEventInfo storedNote in storedNoteEvents)
            {
                if (storedNoteEvents == null) { return; }
                timing = GetTimeDifference(storedNote.startTime);
                if (IsNoteCorrect(note, timing, storedNote))
                {
                    OnNoteSuccess(ref score, note, timing, storedNote);
                    break;  // Exit the loop after the first match

                }

            }

            score ??= GetTimingScore(timing);
            GameManager.instance.combo.ChangeMultiplier(score);

            SpawnPiano.instance.UpdateKeyColours(note - 21, true, score);
            if (GameManager.instance.songTime >= 0)
            {
                GameManager.instance.UpdatePlayerScore(score);
            }
            Replay.UpdateReplay(note, GameManager.instance.songTime);
        }
        enabledKeys[note - 21] = true;



        
    }

    /// <summary>
    /// Handles successful note presses
    /// </summary>
    /// <param name="score"></param>
    /// <param name="note"></param>
    /// <param name="timing"></param>
    /// <param name="storedNote"></param>
    public void OnNoteSuccess(ref string score, int note, float timing, NoteEventInfo storedNote)
    {
        score = GetTimingScore(timing);
        SpawnPiano.instance.SpawnKeyParticle(note - 21, score);
        storedNote.triggered = true;
    }
    /// <summary>
    /// Check if note is was played at a correct timing.
    /// </summary>
    /// <param name="noteNumber">Number of the note played.</param>
    /// <param name="timing"> What time the note was hit.</param>
    /// <param name="storedNote">Note that is being compared to.</param>
    /// <returns></returns>
    public bool IsNoteCorrect(int noteNumber, float timing, NoteEventInfo storedNote)
    {
        if (GameSettings.usePiano) { return noteNumber == storedNote.noteNumber && !storedNote.triggered && timing < 0.5f; }
        return noteNumber % 12 == storedNote.noteNumber % 12 && !storedNote.triggered && timing < 0.5f;

    }


    /// <summary>
    /// Handles MIDI note off events.
    /// </summary>
    /// <param name="channel">The MIDI channel.</param>
    /// <param name="note">The MIDI note number.</param>
    void NoteOff(MidiChannel channel, int note)
    {
        if (!takeInput || isPedalPressed) { return; }
        if (inGame)
        {
            SpawnPiano.instance.UpdateKeyColours(note - 21, false);
            Replay.UpdateReplay(note, GameManager.instance.songTime);
        }
        enabledKeys[note - 21] = false;

    }
    /// <summary>
    /// Reloads the currently playing song.
    /// </summary>
    public void RetryCurrentSong()
    {
        UnHookMidiDevice();
        LoadSongFromCurrentSettings(false);
    }

    /// <summary>
    /// Checks for pressed keys in the 12-key keyboard layout.
    /// </summary>
    void CheckNotesKeyboard12()
    {
        // Check for computer keyboard key presses and releases
        foreach (var keyValuePair in keyboard12)
        {
            KeyCode key = keyValuePair.Key;
            int noteNumber = keyValuePair.Value;

            if (Input.GetKeyDown(key))
            {
                // Simulate Note On event for the pressed key
                NoteOn(MidiChannel.All, noteNumber, 1.0f);
            }

            if (Input.GetKeyUp(key))
            {
                // Simulate Note Off event for the released key
                NoteOff(MidiChannel.All, noteNumber);
            }
        }
    }
    /// <summary>
    /// Checks for pressed keys in the 8-key keyboard layout.
    /// </summary>


    /// <summary>
    /// Checks if any note is currently active (pressed).
    /// </summary>
    /// <returns>True if any note is active, otherwise false.</returns>
    public bool GetAnyNoteActive()
    {
        foreach (bool b in enabledKeys)
        {
            if (b)
            {
                return true;
            }
        }
        return false;
    }

    /// <summary>
    /// Calculates the time difference between the stored timing and the current song time.
    /// </summary>
    /// <param name="storedTiming">The stored timing value.</param>
    /// <returns>The time difference.</returns>
    float GetTimeDifference(float storedTiming)
    {
        // Adjust this threshold based on your timing accuracy requirements
        // float timingThreshold = currentSettings.timeInterval;

        // Check if the received timing is within a certain threshold of the stored timing
        //0.125f leniancy
        return Mathf.Abs( (GameManager.instance.songTime + PlayerSettings.inputDelay)  - 0.125f - storedTiming);
    }
    /// <summary>
    /// Gets the timing score based on the timing difference.
    /// </summary>
    /// <param name="timing">The timing difference.</param>
    /// <returns>The timing score (e.g., "Perfect", "Good", "Okay", "Miss").</returns>
    public string GetTimingScore(float timing)
    {
        if (timing < 0.1f) { return "Perfect"; }
        else if (timing < 0.175f) { return "Good"; }
        else if (timing < 0.275f) { return "Okay"; }
        else { return "Miss"; }
    }
}
